server/src/main/java/org/elasticsearch/common/util/BytesRefHash.java

```diff
@@ -51,6 +55,41 @@ public BytesRefHash(long capacity, float maxLoadFactor, BigArrays bigArrays) {
         spare = new BytesRef();
     }
 
+    public BytesRefHash(BytesRefArray bytesRefArray, BigArrays bigArrays) {
+        this(bytesRefArray, DEFAULT_MAX_LOAD_FACTOR, bigArrays);
+    }
+
+    public BytesRefHash(BytesRefArray byteRefs, float maxLoadFactor, BigArrays bigArrays) {
```

Comment:
It looks like this BytesRefHash constructor will re-allocate twice the hashes and the spare? We also seem to allocate bytesRefs by calling the super constructor, but then we immediately overwrite that reference with the incoming BytesRefArray. I might be confused at what's happening, can you please clarify?

Reply:
it calls `super(...)` -> the constructor of `AbstractHash`, double allocation would happen if calling `this(...)`

Reply:
Ah, I see what I did there :). That makes sense, would it be possible to refactor the code such that we reuse most of the code from `public BytesRefHash(long capacity, float maxLoadFactor, BigArrays bigArrays)`, and then just recompute the hashes?

Reply:
see [commit](https://github.com/elastic/elasticsearch/pull/85826/commits/e9b6b262b5de7da83105811c5e9df712a3afa392), it is a little complex, because we have to ensure we do not leak BytesRefArray.